mlx/backend/metal/ternary.cpp

```diff
@@ -0,0 +1,136 @@
+// Copyright © 2024 Apple Inc.
+
+#include "mlx/backend/common/ternary.h"
+#include "mlx/backend/metal/device.h"
+#include "mlx/backend/metal/utils.h"
+#include "mlx/primitives.h"
+
+#ifndef MLX_METAL_JIT
+#include <fmt/format.h>
+#include "mlx/backend/common/compiled.h"
+#include "mlx/backend/metal/compiled_includes.h"
+#include "mlx/backend/metal/jit/ternary.h"
+#endif
+
+namespace mlx::core {
+
+constexpr int MAX_TERNARY_SPECIALIZED_DIMS = 5;
+
+void ternary_op(
+    const std::vector<array>& inputs,
+    array& out,
+    const std::string op) {
+  assert(inputs.size() == 3);
+  auto& a = inputs[0];
+  auto& b = inputs[1];
+  auto& c = inputs[2];
+  TernaryOpType topt = get_ternary_op_type(a, b, c);
+  set_ternary_op_output_data(a, b, c, out, topt, true /* donate_with_move */);
+
+  if (out.size() == 0) {
+    return;
+  }
+
+  // Try to collapse contiguous dims
+  auto [shape, strides] = collapse_contiguous_dims(a, b, c, out);
+  auto& strides_a = strides[0];
+  auto& strides_b = strides[1];
+  auto& strides_c = strides[2];
+  auto& strides_out = strides[3];
+
+  std::string kernel_name;
+  {
+    std::ostringstream kname;
+    if (topt == TernaryOpType::General) {
+      kname << "g";
+      if (shape.size() <= MAX_TERNARY_SPECIALIZED_DIMS) {
+        kname << shape.size();
+      }
+    } else {
+      kname << "v";
+    }
+    kname << "_" << op << type_to_name(b);
+    kernel_name = kname.str();
+  }
+
+  auto& s = out.primitive().stream();
+  auto& d = metal::device(s.device);
+
+  MTL::ComputePipelineState* kernel;
+
+  if constexpr (mlx_metal_jit()) {
```

Comment:
@jagrit06 @angeloskath  this is an example using the `constexpr` to figure out if we are JITing or not. It's not as messy as I thought it would be provided the right helper utilities.